lib/device.ex

```diff
@@ -0,0 +1,149 @@
+defmodule Onvif.Device do
+  alias Onvif.Device
+  alias Onvif.Discovery.Probe
+
+  @enforce_keys [:username, :password, :ip]
+
+  defstruct @enforce_keys ++
+              [
+                :auth_type,
+                :scopes,
+                :manufacturer,
+                :model,
+                :firmware_version,
+                :serial_number,
+                :hardware_id,
+                :datetime,
+                :ntp,
+                :media_service_path,
+                current_diff: 0,
+                port: 80,
+                supports_media2?: false,
+                device_service_path: "/onvif/device_service",
+                is_nvr?: false
+              ]
+
+  @spec init(Onvif.Discovery.Probe.t(), String.t(), String.t(), boolean, boolean) ::
+          {:error, any()}
+          | %Device{}
+  def init(%Probe{} = probe_match, username, password, use_https? \\ false, use_ipv6? \\ false) do
+    with {:ok, device} <-
+           device_from_probe_match(probe_match, username, password, use_https?, use_ipv6?),
+         {:ok, device_with_datetime} <- get_date_time(device),
+         {:ok, updated_device} <- guess_auth(device_with_datetime) do
+      updated_device
+      |> get_services()
+      |> set_media_service_path()
+    end
+  end
+
+  defp device_from_probe_match(%Probe{} = probe_match, username, password, use_https?, use_ipv6?) do
+    scheme = if use_https?, do: "https", else: "http"
```

Comment:
since scheme is included in the address in the address list, this would probably work better to filter out entire addresses, like the use_ipv6. These should probably say `prefer_https?` and `prefer_ipv6?` because we are bound to what onvif tells us is available. If they don't have any https endpoints we can't just decide to use one and have it work